RandLAPACK/drivers/rl_cqrrpt_gpu.hh

```diff
@@ -0,0 +1,395 @@
+        ///
+        /// The algorithm optionally times all of its subcomponents through a user-defined 'verbosity' parameter.
+        ///
+        /// The algorithm optionally computes a condition number of a preconditioned matrix A through a 'cond_check'
+        /// parameter, which defaults to 0. This requires extra n * (m + 1) * sizeof(T) bytes of space, which will be 
+        /// internally allocated by a utility routine. 
+        /// A computation is handled by a utility method that finds the l2 condition number by computing all singular
+        /// values of the R-factor via an appropriate LAPACK function.
+        CQRRPT_GPU(
+            bool verb,
+            bool time_subroutines,
+            T ep
+        ) {
+            verbosity = verb;
+            timing = time_subroutines;
+            eps = ep;
+            no_hqrrp = 1;
+            nb_alg = 64;
+            oversampling = 10;
+            use_cholqr = 0;
+            panel_pivoting = 1;
+        }
+
+        /// Computes a QR factorization with column pivots of the form:
+        ///     A[:, J] = QR,
+        /// where Q and R are of size m-by-k and k-by-n, with rank(A) = k.
+        /// Detailed description of this algorithm may be found in Section 5.1.2.
+        /// of "the RandLAPACK book". 
+        ///
+        /// @param[in] m
+        ///     The number of rows in the matrix A.
+        ///
+        /// @param[in] n
+        ///     The number of columns in the matrix A.
+        ///
+        /// @param[in] A
+        ///     The m-by-n matrix A, stored in a column-major format.
+        ///
+        /// @param[in] d
+        ///     Embedding dimension of a sketch, m >= d >= n.
+        ///
+        /// @param[in] R
+        ///     Represents the upper-triangular R factor of QR factorization.
+        ///     On entry, is empty and may not have any space allocated for it.
+        ///
+        /// @param[out] A
+        ///     Overwritten by an m-by-k orthogonal Q factor.
+        ///     Matrix is stored explicitly.
+        ///
+        /// @param[out] R
+        ///     Stores k-by-n matrix with upper-triangular R factor.
+        ///     Zero entries are not compressed.
+        ///
+        /// @param[out] J
+        ///     Stores k integer type pivot index extries.
+        ///
+        /// @return = 0: successful exit
+        ///
+        /// @return = 1: cholesky factorization failed
+        ///
+
+        int call(
+            int64_t m,
+            int64_t n,
+            T* A,
+            int64_t lda,
+            T* R,
+            int64_t ldr,
+            int64_t* J,
+            T d_factor,
+            RandBLAS::RNGState<RNG> &state
+        ) override;
+
+    public:
+        bool verbosity;
+        bool timing;
+        T eps;
+        int64_t rank;
+
+        // 8 entries
+        std::vector<long> times;
+
+        // tuning SASOS
+        int num_threads;
+        int64_t nnz;
+
+        // HQRRP-related
+        int no_hqrrp;
```

Comment:
Int seems like a weird type in this context.

Reply:
Why?
CQRRPT is not fully GPU-based, the data gets offloaded onto a GPU only after QRCP on a sketch is done.